add validations and model specs

fickryiman opened the pull request:
- Add the following validations: 
     - For the User model: 
     - Name must not be blank. 
     - PostsCounter must be an integer greater than or equal to zero. 
  - For the Post model: 
     - The title must not be blank. 
     - The title must not exceed 250 characters. 
     - CommentsCounter must be an integer greater than or equal to zero. 
     - LikesCounter must be an integer greater than or equal to zero.
  - Add unit specs for all of the models' methods and validations.

fickryiman committed:
add validations and model specs

  - Add the following validations:
    - For the User model:
      - Name must not be blank.
      - PostsCounter must be an integer greater than or equal to zero.
    - For the Post model:
      - Title must not be blank.
      - Title must not exceed 250 characters.
      - CommentsCounter must be an integer greater than or equal to zero.
      - LikesCounter must be an integer greater than or equal to zero.
  - Add unit specs for all of models' methods and validations.

fickryiman committed:
add validations for user and post models

For the User model:
- Name must not be blank.
- PostsCounter must be an integer greater than or equal to zero.
For the Post model:
- Title must not be blank.
- Title must not exceed 250 characters.
- CommentsCounter must be an integer greater than or equal to zero.
- LikesCounter must be an integer greater than or equal to zero.

fickryiman committed:
model specs unit test prepare

- add Rspec gem in the development
group :development, :test do
  gem 'rspec-rails'
end
- run: `bundle install` in the terminal
- To set up RSpec in the app and create the Spec folder run: `rails generate rspec:install`

fickryiman committed:
run rubocop auto

fickryiman committed:
add rails test

fickryiman committed:
fixes linters error

elisha2kyakpo1 commented on app/models/post.rb:
```diff
@@ -3,6 +3,10 @@ class Post < ApplicationRecord
+  validates :title, presence: true, length: { maximum: 250 }
+  validates :comments_counter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
+  validates :likes_counter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
+
   def recent_comments
     comments.order(created_at: :desc).limit(5)
   end
```
- [x] Kindly test this function and the one that __updates the posts counter__ as it is a requirement to test all the model methods.

elisha2kyakpo1 commented on app/models/user.rb:
```diff
@@ -3,6 +3,9 @@ class User < ApplicationRecord
 
+  validates :name, presence: true
+  validates :posts_counter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
+
   def recent_posts
     posts.order(created_at: :desc).limit(3)
   end
```
- [x] Kindly test this function to see if it can return the 3 most recent posts.

prolajumokeoni replied:
Hi @fickryiman, this test is not yet added 👍🏿

elisha2kyakpo1 commented on spec/like_spec.rb:
```diff
@@ -0,0 +1,13 @@
+require 'rails_helper'
+
+RSpec.describe Like, type: :model do
+  it 'is not valid without an author_id' do
+    like = Like.new(post_id: 1)
+    expect(like).to_not be_valid
+  end
+
+  it 'is not valid without a post_id' do
+    like = Like.new(author_id: 1)
+    expect(like).to_not be_valid
+  end
```
- [x] The function that updates the likes counter is not covered in the tests, kindly test it.

prolajumokeoni replied:
Hi @fickryiman, this test is not yet added 👍🏿

elisha2kyakpo1 reviewed (changes requested):
Hi 👋 @fickryiman, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

✔️ The PR title and description are good and professional.
✔️ No linter errors.
✔️ Descriptive readme file👏
👏 Model validations are set up nicely.

## Required Changes ♻️
_Check the comments under the review._

- [x] The live link in the demo section is not correct, kindly provide a correct link or remove the whole section.

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

fickryiman committed:
add models method test

fickryiman committed:
update readme for validations and unit test project

fickryiman committed:
update readme

frank1738 commented on spec/user_spec.rb:
```diff
@@ -0,0 +1,23 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  it 'is valid with valid attributes' do
+    user = User.new(name: 'user-1', posts_counter: 0)
+    expect(user).to be_valid
+  end
+
+  it 'is not valid without a name' do
+    user = User.new(posts_counter: 0)
+    expect(user).to_not be_valid
+  end
+
+  it 'is not valid with a non-integer posts_counter' do
+    user = User.new(name: 'user-1', posts_counter: 2.5)
+    expect(user).to_not be_valid
+  end
+
+  it 'is not valid with a negative posts_counter' do
+    user = User.new(name: 'user-1', posts_counter: -1)
+    expect(user).to_not be_valid
+  end
+end
```
- [ ] As requested earlier, Kindly test the `recent_posts` method in your user model

frank1738 commented on spec/post_spec.rb:
```diff
@@ -0,0 +1,46 @@
+require 'rails_helper'
+
+RSpec.describe Post, type: :model do
+  it 'is valid with valid attributes' do
+    author = User.create(name: 'author-1')
+    post = Post.new(title: 'post-1', text: 'post text', author:, comments_counter: 0, likes_counter: 0)
+    expect(post).to be_valid
+  end
+
+  it 'is not valid without a title' do
+    author = User.create(name: 'author-1')
+    post = Post.new(text: 'post text', author:, comments_counter: 0, likes_counter: 0)
+    expect(post).to_not be_valid
+  end
+
+  it 'is not valid when a title exceed 250 characters' do
+    author = User.create(name: 'author-1')
+    title = 'a' * 251
+    post = Post.new(title:, text: 'post text', author:, comments_counter: 0, likes_counter: 0)
+    expect(post).to_not be_valid
+  end
```
- [ ] Kindly also test these methods, `recent_comments`  and `update_posts_counter` in your post model to ensure your app behaves as expected

frank1738 reviewed (changes requested):
Hi @fickryiman , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

<img src="https://media1.giphy.com/media/xUOxfh6ZM75efM3Bqo/giphy-downsized-medium.gif"/>


## Highlights

- [ ] Model validations are setup properly


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

N/A
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

fickryiman committed:
update no live demo yet

fickryiman committed:
use test models instead of rspec

fickryiman committed:
add rspec model test

- add user models recent_posts test
- add post models recent_comments test
- add post models update_posts_counter test

aamirkhan2478 reviewed (approved):
Hi @fickryiman, 

![GIF](https://www.funimada.com/assets/images/cards/big/congrats-7.gif)

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
✅ You used Gitflow correctly :+1: 
✅ No linter error :+1: 
✅ Good commit messages :+1: 
✅ Good descriptive PR :+1: 
✅ Well done :clap: you did all changes given by the previous reviewers :+1: 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
